httpcore/_backends/anyio.py

```diff
@@ -0,0 +1,194 @@
+from anyio.streams.tls import TLSAttribute, TLSStream
+
+from .._exceptions import (
+    CloseError,
+    ConnectError,
+    ConnectTimeout,
+    ReadError,
+    ReadTimeout,
+    WriteError,
+    WriteTimeout,
+)
+from .._types import TimeoutDict
+from .base import AsyncBackend, AsyncLock, AsyncSemaphore, AsyncSocketStream
+
+
+class SocketStream(AsyncSocketStream):
+    def __init__(self, stream: ByteStream) -> None:
+        self.stream = stream
+        self.read_lock = anyio.create_lock()
+        self.write_lock = anyio.create_lock()
+
+    def get_http_version(self) -> str:
+        alpn_protocol = self.stream.extra(TLSAttribute.alpn_protocol, None)
+        return "HTTP/2" if alpn_protocol == "h2" else "HTTP/1.1"
+
+    async def start_tls(
+        self,
+        hostname: bytes,
+        ssl_context: SSLContext,
+        timeout: TimeoutDict,
+    ) -> "SocketStream":
+        connect_timeout = timeout.get("connect")
+        try:
+            async with anyio.fail_after(connect_timeout):
+                ssl_stream = await TLSStream.wrap(
+                    self.stream,
+                    ssl_context=ssl_context,
+                    hostname=hostname.decode("ascii"),
+                )
+        except TimeoutError:
+            raise ConnectTimeout from None
+        except BrokenResourceError as exc:
+            raise ConnectError from exc
+
+        return SocketStream(ssl_stream)
+
+    async def read(self, n: int, timeout: TimeoutDict) -> bytes:
+        read_timeout = timeout.get("read")
+        async with self.read_lock:
+            try:
+                async with anyio.fail_after(read_timeout):
+                    return await self.stream.receive(n)
+            except TimeoutError:
+                raise ReadTimeout from None
+            except BrokenResourceError as exc:
+                raise ReadError from exc
+            except EndOfStream:
+                raise ReadError("Server disconnected while attempting read") from None
+
+    async def write(self, data: bytes, timeout: TimeoutDict) -> None:
+        if not data:
+            return
+
+        write_timeout = timeout.get("write")
+        async with self.write_lock:
+            try:
+                async with anyio.fail_after(write_timeout):
+                    return await self.stream.send(data)
+            except TimeoutError:
+                raise WriteTimeout from None
+            except BrokenResourceError as exc:
+                raise WriteError from exc
+
+    async def aclose(self) -> None:
+        async with self.write_lock:
+            try:
+                await self.stream.aclose()
+            except BrokenResourceError as exc:
+                raise CloseError from exc
+
+    def is_connection_dropped(self) -> bool:
+        raw_socket = self.stream.extra(SocketAttribute.raw_socket)
+        rready, _wready, _xready = select.select([raw_socket], [], [], 0)
+        return bool(rready)
```

Comment:
We might want to look into #182 some more before deciding that this trick is worth committing to. 🤔

Reply:
Would it make sense for us to approach it in this order?...

1. Merge this PR.
2. Update #185 to include curio and anyio.
3. Get #185 merged.

Reply:
Sure yup, #185 isn't blocking us from merging this, just wanted to clarify that this select() approach needed a bit of refining. :)